### ISSUE TITLE
Eagerly create RibbonClients

### ISSUE THREAD

**Issue by KamikaZeeFu** [spring-cloud/spring-cloud-netflix#1334]
From what I've understood, each Ribbon Client created by Spring basically gets it's own ApplicationContext.
This context is created when the client is first used, which makes the very first request rather slow.
Is it possible to eagerly create these clients/contexts?

**Comment by spencergibb** [comment]
@KamikaZeeFu not currently. In fact, we've had problems when feign clients (similar to ribbon clients) get eagerly created. Besides the first request is slow, are you experiencing problems?

**Comment by jansyk13** [comment]
@spencergibb Problem is not the lazy initialization, but the fact that the creation of client is not synchronized. ```SpringClientFactory#instantiateWithConfig```

**Comment by spencergibb** [comment]
And what problems does that cause @jansyk13?

**Comment by KamikaZeeFu** [comment]
@spencergibb well the main problem is that it creates unpredictable delays.
I have a setup with a Zuul reverse proxy that calls a couple of backend services.
When i deploy a new version of a backend, the first request will be delayed, which causes random request to run into the zuul configured timeout. My default response time is around 300-400ms, these initial context init spikes push it to above 1s.

Of course I could just increase my timeouts, but that is rather unsatisfying.

**Comment by spencergibb** [comment]
It would likely be feasible to eagerly create the zuul ribbon clients since we know all of the routes ahead of time.

**Comment by jebeaudet** [comment]
In the meantime, you could use something like this to create the clients when the proxy boots, we use that to refresh them if we ever change the read/connect timeout values in archaius : 

```
@Component
public class RibbonEarlyInstantiator implements ApplicationListener<ApplicationStartedEvent>{

    @Autowired
    private ZuulProperties zuulProperties;
    @Autowired
    private SpringClientFactory springClientFactory;

    @Override
    public void onApplicationEvent(ApplicationStartedEvent event)
    {
        Map<String, ZuulRoute> routes = zuulProperties.getRoutes();
        routes.values()
              .stream()
              .filter(route -> route.getServiceId() != null)
              .map(route -> route.getServiceId())
              .distinct()
              .forEach(serviceId -> {
                  logger.info("Instantiating the context for the client '{}'", serviceId);
                  springClientFactory.getClient(serviceId, RibbonLoadBalancingHttpClient.class); //Or RestClient.class if you still use it
              });
    }
}

**Comment by bijukunjummen** [comment]
Upstream issue: https://github.com/spring-cloud/spring-cloud-commons/issues/184

**Comment by spencergibb** [comment]
Closed via 5bf87c4da1e490bdba03891d1e3c0c1bee6992ac
